File: android/app/src/main/kotlin/dev/agentdeck/terrarium/TerrariumState.kt
package dev.agentdeck.terrarium

import dev.agentdeck.net.AgentState
import dev.agentdeck.state.DashboardState
import dev.agentdeck.terrarium.creature.AgentMark

/** Visual states for each creature and the environment. */

enum class OctopusVisualState {
    SLEEPING,    // Curled up at bottom, dim, eyes closed
    FLOATING,    // Gentle sine bob, tentacles wave
    WORKING,     // Starburst animation — processing (tool use or thinking)
    ASKING,      // Speech bubble + "?" — awaiting user input
}

enum class CrayfishVisualState {
    DORMANT,    // Partially hidden behind rocks
    SITTING,    // Idle on rock, claws at rest
    OBSERVING,  // Watching activity — gentle claw fidget, eyes tracking
    ROUTING,    // Claws clap, eyes flash, signal lines emit (OpenClaw orchestrating)
    WAITING,    // Claws raised
    SICK,       // Gateway has errors — desaturated, tilted, labored breathing
}

enum class TetraVisualState {
    ABSENT,     // Not visible (disconnected)
    CIRCLING,   // Boids algorithm, orbiting attractor
    STREAMING,  // Line up, streak horizontally, code trail particles
    HOVERING,   // Formation near options area
}

enum class CloudVisualState {
    DORMANT,     // Faded out — disconnected
    DRIFTING,    // Gentle float — idle
    COMPUTING,   // Pulsing glow — processing
    WAITING,     // Soft blink — awaiting input
}

enum class EnvironmentVisualState {
    DARK,    // Disconnected — dim/off
    CALM,    // Idle — gentle caustics, slow bubbles
    ACTIVE,  // Processing — bright caustics, more bubbles
    ALERT,   // Awaiting input — pulsing highlights
}

/** Per-agent creature state for multi-session rendering. */
data class AgentCreatureState(
    val sessionId: String,
    val agentType: String?,
    val mark: AgentMark?,
    val visualState: OctopusVisualState,
    val isPrimary: Boolean,
    val layoutSlot: Int,
    val displayName: String? = null,
)

/** Combined visual state for the entire terrarium scene. */
data class TerrariumState(
    val octopus: OctopusVisualState,
    val crayfish: CrayfishVisualState,
    val tetra: TetraVisualState,
    val environment: EnvironmentVisualState,
    val currentTool: String? = null,
    val toolProgress: String? = null,
    val projectName: String? = null,
    val modelName: String? = null,
    val agentType: String? = null,
    val hasError: Boolean = false,
    /** Multi-session: all coding agent creatures (octopuses). */
    val agents: List<AgentCreatureState> = emptyList(),
    /** Codex CLI cloud creatures. */
    val cloudCreatures: List<AgentCreatureState> = emptyList(),
    /** OpenCode nested-square creatures. */
    val openCodeCreatures: List<AgentCreatureState> = emptyList(),
    /** OpenClaw backend worker count. */
    val workerCrayfishCount: Int = 0,
    /** Pop burst positions (normalized) — set for 1 frame when ASKING exits. */
    val popBurstPositions: List<Pair<Float, Float>> = emptyList(),
)

/** Derive the most active AgentState from sibling sessions (for daemon mode). */
private fun mostActiveSessionState(siblings: List<dev.agentdeck.net.SessionInfo>): AgentState? {
    // Priority: PROCESSING > AWAITING_* > IDLE > null
    var best: AgentState? = null
    for (s in siblings) {
        if (s.agentType == "daemon") continue
        val mapped = when (s.state) {
            "processing" -> AgentState.PROCESSING
            "awaiting_permission" -> AgentState.AWAITING_PERMISSION
            "awaiting_option" -> AgentState.AWAITING_OPTION
            "awaiting_diff" -> AgentState.AWAITING_DIFF
            "idle" -> AgentState.IDLE
            else -> null
        } ?: continue
        if (best == null || mapped.ordinal > best.ordinal) {
            best = mapped
        }
    }
    return best
}

/** Map DashboardState to visual TerrariumState. */
fun DashboardState.toTerrariumState(): TerrariumState {
    val isOpenClaw = agentType == "openclaw"
    val hasTool = currentTool != null

    // In daemon mode, the daemon's own agentState is DISCONNECTED (no PTY).
    // Derive effective state from the most active sibling session so that
    // tetra/environment/bubbles reflect actual activity.
    val isDaemonLike = agentType == "daemon" ||
        (agentType != null && siblingSessions.any { it.agentType == agentType })
    val effectiveAgentState = if (isDaemonLike && agentState == AgentState.DISCONNECTED) {
        mostActiveSessionState(siblingSessions) ?: AgentState.IDLE
    } else {
        agentState
    }

    val octopus = when (effectiveAgentState) {
        AgentState.DISCONNECTED -> OctopusVisualState.SLEEPING
        AgentState.IDLE -> OctopusVisualState.FLOATING
        AgentState.PROCESSING -> OctopusVisualState.WORKING
        AgentState.AWAITING_PERMISSION,
        AgentState.AWAITING_OPTION,
        AgentState.AWAITING_DIFF -> OctopusVisualState.ASKING
    }

    // OpenClaw sibling state determines crayfish independently
    val ocSibling = siblingSessions.firstOrNull { it.agentType == "openclaw" }
    val crayfish = when {
        // Primary is OpenClaw — use primary state
        isOpenClaw -> when (effectiveAgentState) {
            AgentState.PROCESSING -> CrayfishVisualState.ROUTING
            AgentState.IDLE -> CrayfishVisualState.SITTING
            AgentState.DISCONNECTED -> CrayfishVisualState.DORMANT
            else -> CrayfishVisualState.WAITING
        }
        // Sibling OpenClaw exists — use its state
        ocSibling != null -> when (ocSibling.state) {
            "processing" -> CrayfishVisualState.ROUTING
            "idle" -> CrayfishVisualState.SITTING
            "awaiting_permission", "awaiting_option", "awaiting_diff" -> CrayfishVisualState.WAITING
            else -> if (ocSibling.alive) CrayfishVisualState.SITTING else CrayfishVisualState.DORMANT
        }
        // Gateway detected but no bridge
        gatewayAvailable == true -> CrayfishVisualState.SITTING
        // Nothing — derive from effective agent state
        else -> when (effectiveAgentState) {
            AgentState.DISCONNECTED -> CrayfishVisualState.DORMANT
            AgentState.PROCESSING -> CrayfishVisualState.OBSERVING
            else -> CrayfishVisualState.SITTING
        }
    }

    // Override to SICK if gateway has errors (but not when DORMANT — gateway unreachable)
    val effectiveCrayfish = if (gatewayHasError == true && crayfish != CrayfishVisualState.DORMANT) {
        CrayfishVisualState.SICK
    } else {
        crayfish
    }

    val crayfishRouting = effectiveCrayfish == CrayfishVisualState.ROUTING
    val tetra = when (effectiveAgentState) {
        AgentState.DISCONNECTED -> TetraVisualState.ABSENT
        AgentState.IDLE -> if (crayfishRouting) TetraVisualState.STREAMING else TetraVisualState.CIRCLING
        AgentState.PROCESSING -> if (hasTool || isOpenClaw || crayfishRouting) TetraVisualState.STREAMING else TetraVisualState.CIRCLING
        AgentState.AWAITING_PERMISSION,
        AgentState.AWAITING_OPTION,
        AgentState.AWAITING_DIFF -> TetraVisualState.HOVERING
    }

    val environment = when (effectiveAgentState) {
        AgentState.DISCONNECTED -> EnvironmentVisualState.DARK
        AgentState.IDLE -> EnvironmentVisualState.CALM
        AgentState.PROCESSING -> EnvironmentVisualState.ACTIVE
        AgentState.AWAITING_PERMISSION,
        AgentState.AWAITING_OPTION,
        AgentState.AWAITING_DIFF -> EnvironmentVisualState.ALERT
    }

    // Build multi-agent creature list from sibling sessions
    val agents = mutableListOf<AgentCreatureState>()

    // Primary agent — skip if disconnected (no session), daemon-like, openclaw proxy, codex-cli, or opencode
    if (agentState != AgentState.DISCONNECTED && !isDaemonLike && agentType != "openclaw" && agentType != "codex-cli" && agentType != "opencode") {
        agents.add(
            AgentCreatureState(
                sessionId = sessionId ?: "primary",
                agentType = agentType,
                mark = AgentMark.fromAgentType(agentType),
                visualState = octopus,
                isPrimary = true,
                layoutSlot = 0,
                displayName = projectName,
            )
        )
    }

    // Sibling sessions (coding agents only — not the current session)
    var slot = agents.size
    for (sibling in siblingSessions) {
        if (sessionId != null && sibling.id == sessionId) continue // skip self (null guard)
        val siblingType = sibling.agentType
        if (siblingType == "openclaw" || siblingType == "daemon" || siblingType == "codex-cli" || siblingType == "opencode") continue // not octopus
        agents.add(
            AgentCreatureState(
                sessionId = sibling.id,
                agentType = siblingType,
                mark = AgentMark.fromAgentType(siblingType),
                visualState = mapSessionOctopusState(sibling.state),
                isPrimary = false,
                layoutSlot = slot++,
                displayName = sibling.projectName,
            )
        )
    }

    // Build cloud creatures list from codex-cli sessions
    val cloudCreatures = mutableListOf<AgentCreatureState>()
    // Primary codex-cli agent
    if (agentState != AgentState.DISCONNECTED && !isDaemonLike && agentType == "codex-cli") {
        cloudCreatures.add(
            AgentCreatureState(
                sessionId = sessionId ?: "primary-cloud",
                agentType = agentType,
                mark = AgentMark.fromAgentType(agentType),
                visualState = octopus, // reuse mapped state (will be converted to CloudVisualState at render)
                isPrimary = true,
                layoutSlot = 0,
                displayName = projectName,
            )
        )
    }
    // Sibling codex-cli sessions
    var cloudSlot = cloudCreatures.size
    for (sibling in siblingSessions) {
        if (sessionId != null && sibling.id == sessionId) continue
        if (sibling.agentType != "codex-cli") continue
        cloudCreatures.add(
            AgentCreatureState(
                sessionId = sibling.id,
                agentType = sibling.agentType,
                mark = AgentMark.fromAgentType(sibling.agentType),
                visualState = mapSessionOctopusState(sibling.state),
                isPrimary = false,
                layoutSlot = cloudSlot++,
                displayName = sibling.projectName,
            )
        )
    }

    // Build OpenCode creatures list from opencode sessions
    val openCodeCreatures = mutableListOf<AgentCreatureState>()
    if (agentState != AgentState.DISCONNECTED && !isDaemonLike && agentType == "opencode") {
        openCodeCreatures.add(
            AgentCreatureState(
                sessionId = sessionId ?: "primary-opencode",
                agentType = agentType,
                mark = AgentMark.fromAgentType(agentType),
                visualState = octopus,
                isPrimary = true,
                layoutSlot = 0,
                displayName = projectName,
            )
        )
    }
    var openCodeSlot = openCodeCreatures.size
    for (sibling in siblingSessions) {
        if (sessionId != null && sibling.id == sessionId) continue
        if (sibling.agentType != "opencode") continue
        openCodeCreatures.add(
            AgentCreatureState(
                sessionId = sibling.id,
                agentType = sibling.agentType,
                mark = AgentMark.fromAgentType(sibling.agentType),
                visualState = mapSessionOctopusState(sibling.state),
                isPrimary = false,
                layoutSlot = openCodeSlot++,
                displayName = sibling.projectName,
            )
        )
    }

    // Number duplicate display names: "AgentDeck", "AgentDeck" → "AgentDeck #1", "AgentDeck #2"
    // Group by (displayName, agentType) so different creature types sharing a project name
    // don't get numbered together (matches SessionListPanel + shared session-utils.ts)
    data class NameTypeKey(val name: String?, val type: String?)
    val nameCounts = agents.groupingBy { NameTypeKey(it.displayName, it.agentType) }.eachCount()
    val nameCounters = mutableMapOf<NameTypeKey, Int>()
    for (i in agents.indices) {
        val key = NameTypeKey(agents[i].displayName, agents[i].agentType)
        if (key.name != null && (nameCounts[key] ?: 0) >= 2) {
            val seq = (nameCounters[key] ?: 0) + 1
            nameCounters[key] = seq
            agents[i] = agents[i].copy(displayName = "${key.name} #$seq")
        }
    }

    return TerrariumState(
        octopus = octopus,
        crayfish = effectiveCrayfish,
        tetra = tetra,
        environment = environment,
        currentTool = currentTool,
        toolProgress = toolProgress,
        projectName = projectName,
        modelName = modelName,
        agentType = agentType,
        hasError = gatewayHasError == true,
        agents = agents,
        cloudCreatures = cloudCreatures,
        openCodeCreatures = openCodeCreatures,
        workerCrayfishCount = workerSessionCount ?: 0,
    )
}

private fun mapSessionOctopusState(state: String?): OctopusVisualState = when (state) {
    "processing" -> OctopusVisualState.WORKING
    "awaiting_permission", "awaiting_option", "awaiting_diff" -> OctopusVisualState.ASKING
    "idle" -> OctopusVisualState.FLOATING
    else -> OctopusVisualState.FLOATING
}

File: apple/AgentDeck/Terrarium/TerrariumState.swift
// TerrariumState.swift — Visual state mapping from DashboardState
// Ported from android TerrariumState.kt

import Foundation

// MARK: - Octopus Visual State

enum OctopusVisualState {
    case sleeping   // Curled at bottom, dim
    case floating   // Gentle bob, tentacle wave (idle)
    case working    // Swim + starburst (processing)
    case asking     // Fidget + "?" bubble (awaiting input)
}

// MARK: - Crayfish Visual State

enum CrayfishVisualState {
    case dormant    // Hidden behind rocks
    case sitting    // Idle, heartbeat glow
    case observing  // Watching, gentle fidget
    case routing    // Claws clap, signal waves
    case waiting    // Claws raised
    case sick       // Desaturated, tilted (gateway error)
}

// MARK: - Tetra Visual State

enum TetraVisualState {
    case absent
    case circling   // Boids orbit
    case streaming  // Line up, food chase
    case hovering   // Near options
}

// MARK: - Cloud Creature State

struct CloudCreatureState: Identifiable {
    let id: String
    let projectName: String?
    let modelName: String?
    let state: CloudVisualState
    let homeX: Float
    let homeY: Float
    let scale: Float
    var exitedWaiting = false
}

// MARK: - Agent Creature State

struct AgentCreatureState: Identifiable {
    let id: String           // session ID
    let projectName: String?
    let modelName: String?
    let state: OctopusVisualState
    let homeX: Float
    let homeY: Float
    let scale: Float

    /// Whether this session just exited ASKING (triggers pop burst)
    var exitedAsking = false
}

// MARK: - OpenCode Creature State

struct OpenCodeCreatureState: Identifiable {
    let id: String
    let projectName: String?
    let modelName: String?
    let state: OpenCodeVisualState
    let homeX: Float
    let homeY: Float
    let scale: Float
}

// MARK: - Terrarium State (aggregate)

struct TerrariumState {
    var creatures: [AgentCreatureState] = []
    var cloudCreatures: [CloudCreatureState] = []
    var opencodeCreatures: [OpenCodeCreatureState] = []
    var crayfishState: CrayfishVisualState = .dormant
    var crayfishVisible: Bool = false
    var tetraState: TetraVisualState = .circling
    var environment: EnvironmentVisualState = .calm
    var hasError: Bool = false

    /// Pop burst positions (from ASKING exit)
    var popBurstPositions: [(x: Float, y: Float)] = []
}

// MARK: - Mapping from DashboardState

extension DashboardState {
    func toTerrariumState(previous: TerrariumState? = nil) -> TerrariumState {
        var result = TerrariumState()

        // Primary session creature (skip daemon/openclaw/codex-cli/opencode — they're not octopuses)
        let primaryIsOctopus = state != .disconnected
            && agentType != "daemon"
            && agentType != "openclaw"
            && agentType != "codex-cli"
            && agentType != "opencode"

        // Octopus siblings (exclude daemon/openclaw/codex-cli/opencode), sorted by ID for stable positioning.
        // Also exclude the currently-focused session's id to prevent double-rendering when focus relay
        // promotes a sibling to primary (agentType → claude-code, sessionId → sibling.id).
        let siblings = siblingSessions
            .filter { $0.agentType != "daemon" && $0.agentType != "openclaw" && $0.agentType != "codex-cli" && $0.agentType != "opencode" }
            .filter { !(primaryIsOctopus && $0.id == sessionId) }
            .sorted { $0.id < $1.id }

        let octopusCount = (primaryIsOctopus ? 1 : 0) + siblings.count
        let slots = CreatureLayout.layoutOctopuses(count: octopusCount)

        var creatures: [AgentCreatureState] = []  // mutated later for dedup numbering
        var slotIdx = 0

        if primaryIsOctopus {
            let primaryState = mapToOctopusState(state)
            let slot = slots.first ?? CreatureSlot(x: 0.4, y: 0.45, scale: 1.0)

            // Detect ASKING exit for pop burst
            let wasAsking = previous?.creatures.first?.state == .asking
            let exitedAsking = wasAsking && primaryState != .asking

            creatures.append(AgentCreatureState(
                id: sessionId ?? "primary",
                projectName: projectName,
                modelName: modelName,
                state: primaryState,
                homeX: slot.x,
                homeY: slot.y,
                scale: slot.scale,
                exitedAsking: exitedAsking
            ))
            slotIdx = 1
        }
        // Sibling octopuses
        for (i, sibling) in siblings.enumerated() {
            guard !slots.isEmpty else { break }
            let idx = min(slotIdx + i, slots.count - 1)
            let s = slots[idx]
            let sibState = mapSiblingState(sibling.state)
            creatures.append(AgentCreatureState(
                id: sibling.id,
                projectName: sibling.projectName,
                modelName: sibling.modelName,
                state: sibState,
                homeX: s.x,
                homeY: s.y,
                scale: s.scale
            ))
        }

        // Number duplicate display names: "AgentDeck", "AgentDeck" → "AgentDeck #1", "AgentDeck #2"
        // Group by (projectName, agentType) so different creature types sharing a project name
        // don't get numbered together (matches SessionListPanel + shared session-utils.ts)
        struct NameTypeKey: Hashable { let name: String?; let type: String }
        let creatureType = "claude-code"  // all entries in `creatures` are octopuses
        let nameCounts = Dictionary(grouping: creatures, by: { NameTypeKey(name: $0.projectName, type: creatureType) }).mapValues { $0.count }
        var nameCounters: [String: Int] = [:]
        for i in creatures.indices {
            let key = NameTypeKey(name: creatures[i].projectName, type: creatureType)
            if let name = creatures[i].projectName, (nameCounts[key] ?? 0) >= 2 {
                let seq = (nameCounters[name] ?? 0) + 1
                nameCounters[name] = seq
                creatures[i] = AgentCreatureState(
                    id: creatures[i].id,
                    projectName: "\(name) #\(seq)",
                    modelName: creatures[i].modelName,
                    state: creatures[i].state,
                    homeX: creatures[i].homeX,
                    homeY: creatures[i].homeY,
                    scale: creatures[i].scale,
                    exitedAsking: creatures[i].exitedAsking
                )
            }
        }

        result.creatures = creatures

        // Cloud (Codex CLI sessions)
        let primaryIsCloud = state != .disconnected && agentType == "codex-cli"
        let cloudSiblings = siblingSessions
            .filter { $0.agentType == "codex-cli" }
            .filter { !(primaryIsCloud && $0.id == sessionId) }
            .sorted { $0.id < $1.id }
        let cloudCount = (primaryIsCloud ? 1 : 0) + cloudSiblings.count
        let cloudSlots = CreatureLayout.layoutCloudCreatures(count: cloudCount)

        var cloudCreatures: [CloudCreatureState] = []
        var cloudSlotIdx = 0
        if primaryIsCloud {
            let s = cloudSlots.first ?? CreatureSlot(x: 0.50, y: 0.45, scale: 1.0)
            cloudCreatures.append(CloudCreatureState(
                id: sessionId ?? "cl-primary",
                projectName: projectName,
                modelName: modelName,
                state: mapToCloudState(state),
                homeX: s.x, homeY: s.y, scale: s.scale
            ))
            cloudSlotIdx = 1
        }
        for (i, sibling) in cloudSiblings.enumerated() {
            guard !cloudSlots.isEmpty else { break }
            let idx = min(cloudSlotIdx + i, cloudSlots.count - 1)
            let s = cloudSlots[idx]
            cloudCreatures.append(CloudCreatureState(
                id: sibling.id,
                projectName: sibling.projectName,
                modelName: sibling.modelName,
                state: mapSiblingToCloudState(sibling.state),
                homeX: s.x, homeY: s.y, scale: s.scale
            ))
        }
        result.cloudCreatures = cloudCreatures

        // OpenCode creatures
        let primaryIsOpenCode = state != .disconnected && agentType == "opencode"
        let openCodeSiblings = siblingSessions
            .filter { $0.agentType == "opencode" }
            .filter { !(primaryIsOpenCode && $0.id == sessionId) }
            .sorted { $0.id < $1.id }
        let openCodeCount = (primaryIsOpenCode ? 1 : 0) + openCodeSiblings.count
        let openCodeSlots = CreatureLayout.layoutOpenCodeCreatures(count: openCodeCount)

        var openCodeCreatures: [OpenCodeCreatureState] = []
        var openCodeSlotIdx = 0
        if primaryIsOpenCode {
            let s = openCodeSlots.first ?? CreatureSlot(x: 0.48, y: 0.40, scale: 1.0)
            openCodeCreatures.append(OpenCodeCreatureState(
                id: sessionId ?? "oc-primary",
                projectName: projectName,
                modelName: modelName,
                state: mapToOpenCodeState(state),
                homeX: s.x, homeY: s.y, scale: s.scale
            ))
            openCodeSlotIdx = 1
        }
        for (i, sibling) in openCodeSiblings.enumerated() {
            guard !openCodeSlots.isEmpty else { break }
            let idx = min(openCodeSlotIdx + i, openCodeSlots.count - 1)
            let s = openCodeSlots[idx]
            openCodeCreatures.append(OpenCodeCreatureState(
                id: sibling.id,
                projectName: sibling.projectName,
                modelName: sibling.modelName,
                state: mapSiblingToOpenCodeState(sibling.state),
                homeX: s.x, homeY: s.y, scale: s.scale
            ))
        }
        result.opencodeCreatures = openCodeCreatures

        // Environment state — in daemon mode, derive from most active sibling
        let isDaemonLike = agentType == "daemon" ||
            (agentType != nil && siblingSessions.contains(where: { $0.agentType == agentType }))
        let effectiveState: AgentConnectionState
        if isDaemonLike && state == .disconnected {
            effectiveState = mostActiveSiblingState(siblingSessions) ?? .idle
        } else {
            effectiveState = state
        }
        result.environment = mapToEnvironment(effectiveState)
        result.hasError = gatewayHasError

        // Crayfish (OpenClaw gateway) — visible only when the gateway is
        // authenticated. Reachability alone (`gatewayAvailable`) used to
        // trigger a cheerful crayfish even when the shared token was
        // missing, which misled users into thinking OpenClaw was wired up.
        // Now: no token / no auth → dormant (hidden). Auth error → sick.
        result.crayfishVisible = gatewayConnected || gatewayHasError

        if gatewayHasError {
            result.crayfishState = .sick
        } else if let ocSibling = siblingSessions.first(where: { $0.agentType == "openclaw" }) {
            result.crayfishState = ocSibling.state == "processing" ? .routing : .sitting
        } else if gatewayConnected {
            result.crayfishState = .sitting
        } else {
            result.crayfishState = .dormant
        }

        // Tetra state — use effectiveState so daemon mode isn't stuck on ABSENT
        if effectiveState == .processing || creatures.contains(where: { $0.state == .working }) {
            result.tetraState = .streaming
        } else if effectiveState.isAwaiting {
            result.tetraState = .hovering
        } else if effectiveState == .disconnected {
            result.tetraState = .absent
        } else {
            result.tetraState = .circling
        }

        // Pop bursts from ASKING exits
        result.popBurstPositions = creatures
            .filter { $0.exitedAsking }
            .map { (x: $0.homeX, y: $0.homeY) }

        return result
    }

    private func mapToOctopusState(_ connState: AgentConnectionState) -> OctopusVisualState {
        switch connState {
        case .disconnected: .sleeping
        case .idle: .floating
        case .processing: .working
        case .awaitingPermission, .awaitingOption, .awaitingDiff: .asking
        }
    }

    private func mapSiblingState(_ stateStr: String?) -> OctopusVisualState {
        switch stateStr {
        case "processing": .working
        case "awaiting_permission", "awaiting_option", "awaiting_diff": .asking
        case "idle": .floating
        default: .sleeping
        }
    }

    private func mapToCloudState(_ connState: AgentConnectionState) -> CloudVisualState {
        switch connState {
        case .disconnected: .dormant
        case .idle: .drifting
        case .processing: .pulsing
        case .awaitingPermission, .awaitingOption, .awaitingDiff: .waiting
        }
    }

    private func mapSiblingToCloudState(_ stateStr: String?) -> CloudVisualState {
        switch stateStr {
        case "processing": .pulsing
        case "awaiting_permission", "awaiting_option", "awaiting_diff": .waiting
        case "idle": .drifting
        default: .dormant
        }
    }

    private func mapToOpenCodeState(_ connState: AgentConnectionState) -> OpenCodeVisualState {
        switch connState {
        case .disconnected: .dormant
        case .idle: .drifting
        case .processing: .pulsing
        case .awaitingPermission, .awaitingOption, .awaitingDiff: .waiting
        }
    }

    private func mapSiblingToOpenCodeState(_ stateStr: String?) -> OpenCodeVisualState {
        switch stateStr {
        case "processing": .pulsing
        case "awaiting_permission", "awaiting_option", "awaiting_diff": .waiting
        case "idle": .drifting
        default: .dormant
        }
    }

    private func mapToEnvironment(_ connState: AgentConnectionState) -> EnvironmentVisualState {
        switch connState {
        case .disconnected: .dark
        case .idle: .calm
        case .processing: .active
        case .awaitingPermission, .awaitingOption, .awaitingDiff: .alert
        }
    }

    /// In daemon mode, derive the most active state from sibling sessions.
    private func mostActiveSiblingState(_ siblings: [SessionInfo]) -> AgentConnectionState? {
        func priority(_ s: AgentConnectionState) -> Int {
            switch s {
            case .disconnected: return 0
            case .idle: return 1
            case .processing: return 2
            case .awaitingPermission, .awaitingOption, .awaitingDiff: return 3
            }
        }
        var best: AgentConnectionState? = nil
        for s in siblings {
            if s.agentType == "daemon" { continue }
            let mapped: AgentConnectionState? = switch s.state {
            case "processing": .processing
            case "awaiting_permission": .awaitingPermission
            case "awaiting_option": .awaitingOption
            case "awaiting_diff": .awaitingDiff
            case "idle": .idle
            default: nil
            }
            guard let m = mapped else { continue }
            if best == nil || priority(m) > priority(best!) {
                best = m
            }
        }
        return best
    }
}
